### DISCUSSION TITLE
Why does PowerShell substitute (certain) non-ASCII symbols? Should it?

### DISCUSSION THREAD

**Discussion by roke-julian-lockhart** [PowerShell/PowerShell#24287] [Q&A]
If of relevance, I'm ultimately trying to execute the undermentioned PowerShell code:

> ```PS1
> (snap list) -replace '\s{2,}', "`t" | ConvertFrom-CSV -Delimiter "`t" | Format-Table -AutoSize
> ```

This functions almost perfectly, except that `snap list` produces the undermentioned output:

> ```log
> Name               Version             Rev    Tracking       Publisher              Notes
> bare               1.0                 5      latest/stable  canonical✓             base
> core               16-2.61.4-20240607  17200  latest/stable  canonical✓             core
> core20             20240416            2318   latest/stable  canonical✓             base
> core22             20240823            1612   latest/stable  canonical✓             base
> gtk-common-themes  0.1-81-g442e511     1535   latest/stable  canonical✓             -
> powershell         7.4.5               275    latest/stable  microsoft-powershell✓  classic
> snapd              2.63                21759  latest/stable  canonical✓             snapd
> what-snap          1.0                 8      latest/stable  thomi                  -
> ```

...whereas encapsulating that statement into the aforementioned `(snap list)` produces the undermentioned output:

```log
Name               Version             Rev    Tracking       Publisher               Notes
bare               1.0                 5      latest/stable  canonical**             base
core               16-2.61.4-20240607  17200  latest/stable  canonical**             core
core20             20240416            2318   latest/stable  canonical**             base
core22             20240823            1612   latest/stable  canonical**             base
gtk-common-themes  0.1-81-g442e511     1535   latest/stable  canonical**             -
powershell         7.4.5               275    latest/stable  microsoft-powershell**  classic
snapd              2.63                21759  latest/stable  canonical**             snapd
what-snap          1.0                 8      latest/stable  thomi                   -
```

Notice how `✓` becomes `**`?

**Encodement**

As [this comment explains](https://www.reddit.com/r/PowerShell/comments/1aepwuj/comment/kk9zjma/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button), my encoding preferences might be relevant:

> ```log
> PS /home/RokeJulianLockhart> $OutputEncoding    
> 
> Preamble          : 
> BodyName          : utf-8
> EncodingName      : Unicode (UTF-8)
> HeaderName        : utf-8
> WebName           : utf-8
> WindowsCodePage   : 1200
> IsBrowserDisplay  : True
> IsBrowserSave     : True
> IsMailNewsDisplay : True
> IsMailNewsSave    : True
> IsSingleByte      : False
> EncoderFallback   : System.Text.EncoderReplacementFallback
> DecoderFallback   : System.Text.DecoderReplacementFallback
> IsReadOnly        : True
> CodePage          : 65001
> 
> PS /home/RokeJulianLockhart> [Console]::OutputEncoding                           
> 
> CodePage          : 65001
> IsSingleByte      : False
> EncodingName      : Unicode (UTF-8)
> WebName           : utf-8
> Preamble          : 
> BodyName          : 
> HeaderName        : 
> WindowsCodePage   : 
> IsBrowserDisplay  : 
> IsBrowserSave     : 
> IsMailNewsDisplay : 
> IsMailNewsSave    : 
> EncoderFallback   : System.Text.EncoderReplacementFallback
> DecoderFallback   : System.Text.DecoderReplacementFallback
> IsReadOnly        : False
> ```

Summarily, I'm using UTF-8.

**Comment by rhubarb-geek-nz** [comment]
I presume snap list is a Linux command, say on Ubuntu. When you say "produces the undermentioned output:" is that when writing directly to the terminal with no PowerShell filtering?

Do you get the same ✓ to ** conversion just doing....

```(snap list)  | ConvertFrom-CSV | Format-Table -AutoSize```

One explanation may be that "snap list" is not generating UTF-8 in its output, but in the character set for console.

**Reply by roke-julian-lockhart** [reply]
> When you say "produces the undermentioned output:" is that when writing directly to the terminal with no PowerShell filtering?

@rhubarb-geek-nz, yes. Specifically, merely `(snap list)` causes the tick to be substituted with `**`, whereas solely `snap list` appears to function correctly.

> Do you get the same ✓ to ** conversion just doing....
> 
> ```PS1
> (snap list)  | ConvertFrom-CSV | Format-Table -AutoSize
> ```

I'm confident I do, but I'll verify when I'm next at my PC.

> One explanation may be that "snap list" is not generating UTF-8 in its output, but in the character set for console.

Can you elaborate on this? Is it rather like https://codeberg.org/smxi/inxi/issues/280#issue-479535, where the terminal uses such prefixes to determine colour coding?

I wasn't aware that such instructions were advanced enough to perform character substitutions. Irrespective, in that case, I would have expected to see something like `%370` like you would in a URI, not two stars (for surely such arbitrary substitutions wouldn't scale).

**Reply by rhubarb-geek-nz** [reply]
> Specifically, merely `(snap list)` causes the tick to be substituted with `**`, whereas solely `snap list` appears to function correctly.

I expect that when you run "snap list" the output goes directly to the terminal. When you bracket it then the output goes through the pipeline where it gets converted to a UNICODE dotnet string .

What does `locale` say? Normally programs use LANG, LC_C* or LC_ALL to determine character set for output. Looks like a mismatch between the output of `snap list` and pipeline string processing.

**Reply by roke-julian-lockhart** [reply]
> https://github.com/PowerShell/PowerShell/discussions/24287#discussioncomment-10680487

@rhubarb-geek-nz:

<blockQuote>

```log
PS /home/RokeJulianLockhart> locale | ConvertFrom-StringData | Format-Table -AutoSize

Name              Value
----              -----
LANG              en_GB.UTF-8
LC_CTYPE          "en_GB.UTF-8"
LC_NUMERIC        C
LC_TIME           "en_GB.UTF-8"
LC_COLLATE        "en_GB.UTF-8"
LC_MONETARY       C
LC_MESSAGES       "en_GB.UTF-8"
LC_PAPER          C
LC_NAME           C
LC_ADDRESS        C
LC_TELEPHONE      C
LC_MEASUREMENT    C
LC_IDENTIFICATION "en_GB.UTF-8"
LC_ALL
```

</blockQuote>

The ones set to `C` are deliberately set so, because they force the DE to use more international formats, like ISO 8601, etc. However, they're for the more niche purposes (like `LC_TELEPHONE`) so I would be surprised if they affected this.

**Comment by rhubarb-geek-nz** [comment]
It is from the output of snap list. The output varies depending on if the output is a tty or a pipe/file.

Eg, without powershell, just from bash try


```
$ snap list | od -x | grep 2a
```

And you should see a number of 2a values which are the asterisks.

**Reply by roke-julian-lockhart** [reply]
> https://github.com/PowerShell/PowerShell/discussions/24287#discussioncomment-10680733

@rhubarb-geek-nz, indeed:

> ```log
> PS /home/RokeJulianLockhart> bash                          
> RokeJulianLockhart@sayw4i:~$ snap list | od -x | grep -e '2a'
> 0000240 2a6c 202a 2020 2020 2020 2020 2020 2020
> 0000360 2020 6163 6f6e 696e 6163 2a6c 202a 2020
> 0000520 696e 6163 2a6c 202a 2020 2020 2020 2020
> 0000640 6261 656c 2020 6163 6f6e 696e 6163 2a6c
> 0000660 202a 2020 2020 2020 2020 2020 2020 6162
> 0001000 6163 6f6e 696e 6163 2a6c 202a 2020 2020
> 0001140 702d 776f 7265 6873 6c65 2a6c 202a 6320
> 0001260 656c 2020 6163 6f6e 696e 6163 2a6c 202a
> ```

If of use (to compare) `snap list | od -x` does appear to provide identical output in `pwsh`:

<blockQuote>

```log
PS /home/RokeJulianLockhart> snap list | od -x | Select-String -Pattern '2a' | Select @('LineNumber', 'Line', 'Path')

LineNumber Line                                            Path
---------- ----                                            ----
        11 0000240 2a6c 202a 2020 2020 2020 2020 2020 2020 InputStream
        16 0000360 2020 6163 6f6e 696e 6163 2a6c 202a 2020 InputStream
        22 0000520 696e 6163 2a6c 202a 2020 2020 2020 2020 InputStream
        27 0000640 6261 656c 2020 6163 6f6e 696e 6163 2a6c InputStream
        28 0000660 202a 2020 2020 2020 2020 2020 2020 6162 InputStream
        33 0001000 6163 6f6e 696e 6163 2a6c 202a 2020 2020 InputStream
        39 0001140 702d 776f 7265 6873 6c65 2a6c 202a 6320 InputStream
        44 0001260 656c 2020 6163 6f6e 696e 6163 2a6c 202a InputStream
```

</blockQuote>

**Comment by rhubarb-geek-nz** [comment]
… that said; it does encode ellipsis as E2 80 A6 which translates as u+2026. So the output of snap list to a pipe is still UTF8.

**Reply by rhubarb-geek-nz** [reply]
Also try

```
$ NO_COLOR=1 snap list
```

and the tick is now monochrome

**Reply by roke-julian-lockhart** [reply]
> [and the tick is now monochrome](https://github.com/PowerShell/PowerShell/discussions/24287#discussioncomment-10684173)

@rhubarb-geek-nz, not for me:

<blockQuote>

```HTML
<span style="font-family:monospace"><span style="color:#fcfcfc;background-color:#232627;">Script started, output log file is '/home/RokeJulianLockhart/output.txt.log', input log file is '/home/RokeJulianLockhart/input.txt.log', timing file is '/home/RokeJulianLockhart/timings.txt.log'.
</span><br>PowerShell 7.4.5
<br>PS /home/RokeJulianLockhart&gt; <span style="color:#fdbc4b;background-color:#232627;">bash</span><span style="color:#fcfcfc;background-color:#232627;">
</span><br><span style="color:#11d116;background-color:#232627;">RokeJulianLockhart@sayw4i</span><span style="color:#fcfcfc;background-color:#232627;">:</span><span style="color:#11d116;background-color:#232627;">~</span><span style="color:#fcfcfc;background-color:#232627;">$ NO_COLOR=1 snap list
</span><br>Name &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;Version &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;Rev &#160;&#160;&#160;Tracking &#160;&#160;&#160;&#160;&#160;&#160;Publisher &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;Notes
<br>bare &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;1.0 &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;5 &#160;&#160;&#160;&#160;&#160;latest/stable &#160;canonical<span style="font-weight:bold;color:#3daee9;background-color:#232627;">✓</span><span style="color:#fcfcfc;background-color:#232627;"> &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;base
</span><br>core &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;16-2.61.4-20240607 &#160;17200 &#160;latest/stable &#160;canonical<span style="font-weight:bold;color:#3daee9;background-color:#232627;">✓</span><span style="color:#fcfcfc;background-color:#232627;"> &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;core
</span><br>core20 &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;20240705 &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;2379 &#160;&#160;latest/stable &#160;canonical<span style="font-weight:bold;color:#3daee9;background-color:#232627;">✓</span><span style="color:#fcfcfc;background-color:#232627;"> &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;base
</span><br>core22 &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;20240823 &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;1612 &#160;&#160;latest/stable &#160;canonical<span style="font-weight:bold;color:#3daee9;background-color:#232627;">✓</span><span style="color:#fcfcfc;background-color:#232627;"> &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;base
</span><br>gtk-common-themes &#160;0.1-81-g442e511 &#160;&#160;&#160;&#160;1535 &#160;&#160;latest/stable &#160;canonical<span style="font-weight:bold;color:#3daee9;background-color:#232627;">✓</span><span style="color:#fcfcfc;background-color:#232627;"> &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;-
</span><br>powershell &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;7.4.5 &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;275 &#160;&#160;&#160;latest/stable &#160;microsoft-powershell<span style="font-weight:bold;color:#3daee9;background-color:#232627;">✓</span><span style="color:#fcfcfc;background-color:#232627;"> &#160;classic
</span><br>snapd &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;2.63 &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;21759 &#160;latest/stable &#160;canonical<span style="font-weight:bold;color:#3daee9;background-color:#232627;">✓</span><span style="color:#fcfcfc;background-color:#232627;"> &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;snapd
</span><br>what-snap &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;1.0 &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;8 &#160;&#160;&#160;&#160;&#160;latest/stable &#160;thomi &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;-
<br><span style="color:#11d116;background-color:#232627;">RokeJulianLockhart@sayw4i</span><span style="color:#fcfcfc;background-color:#232627;">:</span><span style="color:#11d116;background-color:#232627;">~</span><span style="color:#fcfcfc;background-color:#232627;">$ &#160;</span><br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br><br>
<br></span>
```

</blockQuote>

**Comment by rhubarb-geek-nz** [comment]
In summary it looks like "snap list" is just doing it's thing. The replacement of a tick with a pair of asterisks is not down to PowerShell but by any non-tty stdout (pipe, file).

**Reply by roke-julian-lockhart** [reply]
> https://github.com/PowerShell/PowerShell/discussions/24287#discussioncomment-10684235

@rhubarb-geek-nz, so it's `snap` which is replacing substituting the symbols? In that case, I'll file an issue with `snapd`. Thanks.
